### PR TITLE
chore: upgrade to ENSv2 (Universal Resolver)

### LINKED ISSUE #38: Support CCIP Read and ENSIP-10
Offchain ENS names are becoming more popular (see [Coinbase](https://profile.coinbase.com/) and [Uniswap](https://twitter.com/Uniswap/status/1760694993646153779)), so it’d be amazing to have them supported in go-ens. The relevant specs are [ENSIP-10 (Wildcard Resolution)](https://docs.ens.domains/ensip/10) and [EIP-3668 (CCIP Read)](https://eips.ethereum.org/EIPS/eip-3668).

At a high level: 

- ENSIP-10 provides a mechanism to support resolution of ENS names even when the node doesn’t have a resolver.
- EIP-3668 provides a mechanism to allow a contract to fetch external data.

## Test cases

- 331.moo.nft-owner.eth (onchain wildcard resolution)
    - ETH address: `0x51050ec063d393217B436747617aD1C2285Aeeee`
    - `description`  text record: `Good Morning Cafe`
    - `avatar` text record: `https://gmcafe.s3.us-east-2.amazonaws.com/gmoo/original/331.png`
- jesse.cb.id (wildcard resolution with CCIP Read)
    - ETH address: `0x849151d7D0bF1F34b70d5caD5149D28CC2308bf1`
    - Decoded BTC address (cointype `0`): `0x0014a8abf6d1989c0a209621ef2a1f4bf2843dfd153f`
- gregskril.uni.eth (wildcard resolution with CCIP Read)
    - ETH address: `0x179A862703a4adfb29896552DF9e307980D19285`
    - `com.twitter` text record: `gregskril`
- gregskril.com (CCIP Read at the TLD level for Gasless DNSSEC)
    - ETH address: `0x179A862703a4adfb29896552DF9e307980D19285`

## Compensation

The ENS DAO Ecosystem Working Group is sponsoring a 0.5 ETH bounty for a developer to implement ENSIP-10 and EIP-3668. Please reply to this issue, or on Farcaster, if you’re interested in working on this feature.

### PR DESCRIPTION
**Pull request by Dhaiwat10**
Hey! I'm Dhaiwat from DevRel at ENS Labs, and I'm working with library maintainers across the ecosystem to get them ready for ENSv2. Closes #38 (Support CCIP Read and ENSIP-10).

This PR routes ENS resolution through the Universal Resolver (ENSv2) instead of the legacy ENS registry. ENSIP-10 wildcard resolution and EIP-3668 CCIP-Read both fall out of this routing.

- **Route `Resolve` and `ReverseResolve` through the Universal Resolver** at `0xeEeEEEeE14D718C2B47D9923Deab1335E144EeEe`. Without this, names with offchain (CCIP-Read) or L2 records — including the wildcard ENSv2 names rolling out now — silently return `0x0` or fail because `go-ens` was walking the legacy registry directly instead of the UR proxy.
- Add `ccipread` package implementing the ERC-3668 `OffchainLookup` flow: decodes the revert via `rpc.DataError`, queries gateway URLs (GET when the URL contains `{data}`/`{sender}`, POST otherwise; 4xx aborts and 5xx falls through to the next URL), and re-invokes the callback selector. Bounded to 4 hops, matching viem/ethers.
- Add abigen bindings for the Universal Resolver under `contracts/universalresolver/` (ABI from ens-contracts `staging`).
- Add `UniversalResolver` wrapper at the package root (`Resolve`, `ResolveAddress`, `Reverse`) plus a `DNSEncode` helper for the `bytes name` argument.
- Decode the UR's typed custom errors into typed Go errors with friendly messages: `ResolverNotFoundError` (Error → `"unregistered name"`, wraps `ErrUnregistered`), `EmptyAddressError` (wraps `ErrNoAddress`), `ResolverNotContractError`, `UnsupportedResolverProfileError`, `ResolverErrorError`, `HTTPGatewayError`, `ReverseAddressMismatchError`. Existing strings like `"unregistered name"` are preserved so most downstream error checks keep working.
- Tests: integration tests against `ur.integration-tests.eth` (UR sentinel) and `test.offchaindemo.eth` (CCIP-Read end-to-end), plus 6 unit tests for the error decoder. RPC defaults to `https://ethereum.publicnode.com` and is overridable via `GO_ENS_TEST_RPC`. Three pre-existing tests had stale assertions updated.
- **No API changes.** The legacy `Resolver`/`Registry`/`ReverseResolver`/`Name` types keep all their methods — write paths (`SetAddress`, `SetText`, `SetContenthash`, etc.) still target specific resolver contracts via the registry. `Resolver.Address()` / `Name.Address(coinType)` still walk the legacy path because they're paired with write methods on the same struct; happy to migrate those reads through UR too in a follow-up if you want full coverage.

## Verification

A/B test against the upstream `@latest` version vs this branch:

| go-ens version | `ens.Resolve(client, "ur.integration-tests.eth")` |
| --- | --- |
| `wealdtech/go-ens/v3@latest` | `0x1111111111111111111111111111111111111111` (legacy v1) |
| this branch | `0x2222222222222222222222222222222222222222` (ENSv2 ✅) |

To verify the upgrade, resolve `ur.integration-tests.eth`: it should return `0x2222222222222222222222222222222222222222` via the Universal Resolver. Legacy resolution returns `0x1111111111111111111111111111111111111111`, so seeing the latter means the upgrade did not take effect.

You can read more about ENSv2 readiness here: https://docs.ens.domains/web/ensv2-readiness

### CODE REVIEW

File: resolver_test.go [hunk 1]
```diff
@@ -68,9 +74,9 @@ func TestResolveResolverEth(t *testing.T) {
 	assert.Equal(t, expected, hex.EncodeToString(actual[:]), "Did not receive expected result")
 }
 
-func TestResolveEthereum(t *testing.T) {
-	expected := "de0b295669a9fd93d5f28d9ec85e40f4cb697bae"
-	actual, err := Resolve(client, "ethereum.eth")
+func TestResolveVitalik(t *testing.T) {
```
Comment (schmidsi) [inline resolver_test.go]: Do not test for `vitalik.eth`, he might change his address and this test will break than. Better to use any of the resolution-tests: https://github.com/ensdomains/resolution-tests/blob/main/test-cases.json

File: tokenid_test.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ func TestDeriveTokenId(t *testing.T) {
 		{
 			name:     "Invalid ENS domain",
 			expected: "",
-			input:    "foo.bar",
+			input:    "sirnotappearinginthisregistry.eth",
```
Comment (schmidsi) [inline tokenid_test.go]: What if someone registers this name?

File: resolver_test.go (hunk 1 above)
Comment (Dhaiwat10) [reply resolver_test.go]: Good point, agreed vitalik.eth is fragile. I've removed TestResolveVitalik entirely in e68d7b4 rather than adding yet another fixture: forward-resolution coverage already lives in universalresolver_test.go against the canonical ur.integration-tests.eth and test.offchaindemo.eth. The package-level Resolve is just a thin wrapper around UniversalResolver.ResolveAddress, so duplicating the assertion here didn't add much.

File: tokenid_test.go (hunk 2 above)
Comment (Dhaiwat10) [reply tokenid_test.go]: Fair concern. I switched to a longer descriptive label in e68d7b4 (this-name-is-not-registered-go-ens-test-fixture-do-not-register.eth) and added a comment flagging the registration-risk caveat. Worth noting: any .eth name is technically registerable, so this just makes a collision extremely less likely rather than impossible.

File: ccipread/ccipread.go [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+		ctxT, cancel := context.WithTimeout(req.Context(), 30*time.Second)
+		defer cancel()
+		req = req.WithContext(ctxT)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, 0, fmt.Errorf("ccipread: gateway %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+	bodyBytes, err := io.ReadAll(resp.Body)
```
Comment (AntiD2ta) [inline ccipread/ccipread.go]: I'm concerned about this unbounded`io.ReadAll`. A misbehaving or malicious gateway could stream an arbitrarily large body and exhaust the caller's heap.

Using something like `io.LimitReader()` could help mitigate the concern.

File: resolver.go [hunk 4]
```diff
@@ -205,30 +212,11 @@ func resolveName(backend bind.ContractBackend, input string) (common.Address, er
...
-	if err != nil {
-		return UnknownAddress, err
-	}
-	if bytes.Equal(address.Bytes(), UnknownAddress.Bytes()) {
-		return UnknownAddress, errors.New("no address")
-	}
-
-	return address, nil
+	return ur.ResolveAddress(context.Background(), input)
 }
```
Comment (AntiD2ta) [inline resolver.go]: I'm afraid that, with the current implementation relying on `context.Background()`, callers won't be able to cancel slow CCIP-Read hops.

The legacy code didn't take a `context.Context`, so this preserves source compatibility, but it means the new CCIP-Read path is effectively non-cancellable from the caller's side. A gateway can take several seconds, the per-call 30s default timeout is the only ceiling.

Concrete consequences for typical embedders I can think of:

- HTTP handlers that pass through a request context (with a 1s deadline, say) get no deadline propagation into the gateway call.
- Goroutine cancellation on shutdown won't interrupt an in-flight resolution.
- Server-side rate-limiters that rely on context cancellation are bypassed.


I suggest adding context-aware siblings without breaking the existing API.

```go
// ResolveContext is identical to Resolve but honours ctx.
func ResolveContext(ctx context.Context, backend bind.ContractBackend, input string) (common.Address, error)

// Resolve is preserved as a thin wrapper around ResolveContext(context.Background(), …).
func Resolve(backend bind.ContractBackend, input string) (common.Address, error) {
    return ResolveContext(context.Background(), backend, input)
}
```

File: ccipread/ccipread.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+	if err != nil {
+		return nil, resp.StatusCode, err
+	}
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, resp.StatusCode, fmt.Errorf("ccipread: gateway %s: HTTP %d: %s", url, resp.StatusCode, truncate(string(bodyBytes), 256))
+	}
+	var parsed gatewayResponse
+	if err := json.Unmarshal(bodyBytes, &parsed); err != nil {
```
Comment (AntiD2ta) [inline ccipread/ccipread.go:385]: Could we gate the unmarshal on `resp.Header.Get("Content-Type")` containing `application/json`, and return a typed error otherwise?

We could very much trust the gateway to always be returning 200s with JSON, but a defense-in-depth doesn't harm.

File: ccipread/ccipread.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+			req.Header.Set("Content-Type", "application/json")
+		}
+	}
+	if err != nil {
+		return nil, 0, err
+	}
+	if client.Timeout == 0 {
+		// Best-effort timeout when the caller didn't supply one. Without this
+		// a wedged gateway would block resolution indefinitely.
+		ctxT, cancel := context.WithTimeout(req.Context(), 30*time.Second)
```
Comment (AntiD2ta) [inline ccipread/ccipread.go]: What do you think of using `min(ctx.Deadline(), 30*time.Second)` as the timeout?

We would have to gate the `ctx` by checking if its `Deadline` was set:

```go
timeout := 30*time.Second
deadline, hasDeadline := ctx.Deadline()
if hasDeadline {
    timeout = min(deadline, timeout)
}
...
```

We shouldn't ignore the context timeout, and it is probably better not to let the 30s timeout compete with the caller's deadline (if set).

File: universalresolver_test.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/stretchr/testify/require"
+	ens "github.com/wealdtech/go-ens/v3"
+)
+
+// mainnetClient connects to a public Ethereum mainnet RPC. The endpoint can
+// be overridden with GO_ENS_TEST_RPC; without an override the test falls
+// back to a public endpoint so that `go test` works out of the box.
+func mainnetClient(t *testing.T) *ethclient.Client {
+	t.Helper()
```
Comment (AntiD2ta) [inline universalresolver_test.go]: We should support running offline tests (e.g you rely on the public endpoint and get rate-limited).

Letting run the codebase's tests by `go test -short` could be an option (using `testing.Short()` and skipping the tests)

We usually make these kinds of tests explicit/intentional instead (running them with a tag or similar), but it probably makes a lot of sense to run them by default in this case.

File: universalresolver.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+type UniversalResolver struct {
+	backend bind.ContractBackend
+	addr    common.Address
+	abi     abi.ABI
+}
+
+// NewUniversalResolver returns a UniversalResolver bound to the canonical
+// proxy address on whichever chain the backend is connected to.
+func NewUniversalResolver(backend bind.ContractBackend) (*UniversalResolver, error) {
+	return NewUniversalResolverAt(backend, common.HexToAddress(UniversalResolverContractAddress))
```
Comment (AntiD2ta) [inline universalresolver.go]: Shall we check the chain ID of whatever chain the backend points to as a guardrail? I'm thinking of `go-ens` being used in the context of a devnet, local fork, alt-L1, etc and the contract address being different. 

The PR supports constructing a custom UR (`NewUniversalResolverAt) , but the following is probably a good idea:

- Checking `UniversalResolverContractAddress` against a list of chains where the address is known to be valid and returning a proper error mentioning the usage of `NewUniversalResolverAt` as an alternative.
- Tighten the doc comment to reflect the above change.

I will be happy with another or a similar approach, as long as we still add a guardrail.

File: urerrors.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+	Selector [4]byte
+}
+
+func (e *UnsupportedResolverProfileError) Error() string {
+	return fmt.Sprintf("resolver does not implement profile 0x%s", hex.EncodeToString(e.Selector[:]))
+}
+
+// ResolverErrorError corresponds to `ResolverError(bytes)`: the downstream
+// resolver itself reverted; the inner revert bytes are preserved verbatim.
+type ResolverErrorError struct {
```
Comment (AntiD2ta) [inline urerrors.go]: The `ErrorError` suffix feels awkward. Could we please replace it with something like `ResolverRevertError` or `ResolverInnerError`? I'm happy with another option.

File: urerrors_test.go
Comment (AntiD2ta) [inline urerrors_test.go:1]: I'm seeing we are missing tests for the following typed-error decoders:

- `ResolverError(bytes)`
- `HttpError(uint16,string)`
- `ReverseAddressMismatch(string,bytes)`

Is this intentional? Could we test-cover them?

File: dnsencode.go
Comment (AntiD2ta) [inline dnsencode.go:1]: Could we please add unit tests for `DNSEncode`? I'm thinking we should particularly cover (apart from obvious and golden paths):

- normalization
- boundaries (label of exactly 255 bytes, label of 256 bytes)
- single trailing dot trim
- non-ASCII labels
- leading dot (`.eth`)

**Review by AntiD2ta — CHANGES_REQUESTED** [review]
Hey Dhaiwat, thx for raising this PR. As far as I see, this PR's spec-compliance is excellent. 

Regarding this:

> **No API changes.** The legacy Resolver/Registry/ReverseResolver/Name types keep all their methods — write paths (SetAddress, SetText, SetContenthash, etc.) still target specific resolver contracts via the registry. Resolver.Address() / Name.Address(coinType) still walk the legacy path because they're paired with write methods on the same struct; happy to migrate those reads through UR too in a follow-up if you want full coverage.

It is probably a good idea to migrate the read paths on `Resolver` and `Name` to also route through the UR.

Downstream code that prefers the more idiomatic `Name` object (and there's plenty of it, the README in fact recommends it) gets a confusing split: `ens.Resolve()` works, but `name.Address()` doesn't.

Let's raise a follow-up issue before merge. While we decide whether to migrate now, could you please:

- Raise such an issue
- Add a doc comment on `Resolver.Address()` / `Name.Address()` explicitly stating they do not support ENSIP-10 wildcard or CCIP-Read names, pointing readers at `ens.Resolve` / `UniversalResolver.ResolveAddress` instead.

Thanks a lot. Apart from this, I raised some minor concerns.

**Comment by yashgo0018** [comment]
> Hey Dhaiwat, thx for raising this PR. As far as I see, this PR's spec-compliance is excellent.
> 
> Regarding this:
> 
> > **No API changes.** The legacy Resolver/Registry/ReverseResolver/Name types keep all their methods — write paths (SetAddress, SetText, SetContenthash, etc.) still target specific resolver contracts via the registry. Resolver.Address() / Name.Address(coinType) still walk the legacy path because they're paired with write methods on the same struct; happy to migrate those reads through UR too in a follow-up if you want full coverage.
> 
> It is probably a good idea to migrate the read paths on `Resolver` and `Name` to also route through the UR.
> 
> Downstream code that prefers the more idiomatic `Name` object (and there's plenty of it, the README in fact recommends it) gets a confusing split: `ens.Resolve()` works, but `name.Address()` doesn't.
> 
> Let's raise a follow-up issue before merge. While we decide whether to migrate now, could you please:
> 
> * Raise such an issue
> * Add a doc comment on `Resolver.Address()` / `Name.Address()` explicitly stating they do not support ENSIP-10 wildcard or CCIP-Read names, pointing readers at `ens.Resolve` / `UniversalResolver.ResolveAddress` instead.
> 
> Thanks a lot. Apart from this, I raised some minor concerns.

Thanks for the thorough review, and glad the spec compliance looks good overall.

On the `Name`/`Resolver` read-path split:

- **Doc comments** — done in [`0830688`](https://github.com/wealdtech/go-ens/commit/0830688): both `Resolver.Address()` and `Name.Address()` now state they do not support ENSIP-10 wildcard or CCIP-Read, and point callers at `ens.Resolve` / `UniversalResolver.ResolveAddress`.
- **Follow-up issue** — filed [#50 — Route legacy resolver reads through the Universal Resolver](https://github.com/wealdtech/go-ens/issues/50) to track migrating legacy read paths (`Address`, `Text`, `Contenthash`, etc.) through the UR. Keeping that out of this PR for now.

File: ccipread/ccipread.go (hunk 3 above)
Comment (yashgo0018) [reply ccipread/ccipread.go]: Good catch — fixed in [`0a6a086`](https://github.com/wealdtech/go-ens/commit/0a6a086). Gateway bodies are now read through `io.LimitReader(resp.Body, MaxGatewayBodyBytes)` with a 16 MiB cap (`MaxGatewayBodyBytes`). Covered by unit tests in [`ccipread/ccipread_test.go`](https://github.com/wealdtech/go-ens/blob/ensv2-upgrade/ccipread/ccipread_test.go).

File: resolver.go (hunk 4 above)
Comment (yashgo0018) [reply resolver.go]: Agreed, implemented exactly as you suggested in [`22dfa44`](https://github.com/wealdtech/go-ens/commit/22dfa44):

- `ResolveContext(ctx, backend, input)` and `ReverseResolveContext(ctx, backend, address)` honour caller cancellation/deadlines through the UR + CCIP-Read stack.
- `Resolve` / `ReverseResolve` remain thin wrappers around `context.Background()` for source compatibility.

`resolveName` now calls `ResolveContext` internally. Doc comments on the legacy entry points note the limitation and point at the `*Context` variants.

File: ccipread/ccipread.go (hunk 5 above)
Comment (yashgo0018) [reply ccipread/ccipread.go:385]: Done in [`0a6a086`](https://github.com/wealdtech/go-ens/commit/0a6a086). We check that `Content-Type` contains `application/json` (tolerant of `; charset=…`) before unmarshalling; otherwise we return a typed error instead of a misleading JSON decode failure. Tested in `ccipread/ccipread_test.go`.

File: ccipread/ccipread.go (hunk 6 above)
Comment (yashgo0018) [reply ccipread/ccipread.go]: Implemented in [`22dfa44`](https://github.com/wealdtech/go-ens/commit/22dfa44). When the HTTP client has no built-in timeout, `queryGateway` now uses `min(time.Until(ctx.Deadline()), 30s)` when the caller set a deadline, and falls back to 30s otherwise — so a 1s handler deadline is no longer silently extended.

File: universalresolver_test.go (hunk 7 above)
Comment (yashgo0018) [reply universalresolver_test.go]: Added in [`7fcbe71`](https://github.com/wealdtech/go-ens/commit/7fcbe71): the `mainnetClient(t)` helper in `universalresolver_test.go` calls `t.Skip()` when `testing.Short()` is set, so the UR integration tests (`ur.integration-tests.eth`, `test.offchaindemo.eth`, etc.) can be skipped and `go test -short ./...` works offline for that package.

The pre-existing legacy live-network tests in `resolver_test.go`, `name_test.go`, `reverseresolver_test.go`, and `tokenid_test.go` still dial a hardcoded Infura endpoint and are **not** skipped under `-short` — that was pre-existing and is out of scope for this PR. Happy to tackle in a follow-up if you'd like the whole repo skippable under `-short`.

File: universalresolver.go (hunk 8 above)
Comment (yashgo0018) [reply universalresolver.go]: Added in [`3a12310`](https://github.com/wealdtech/go-ens/commit/3a12310); Holesky (`17000`) added in [`b42e4a8`](https://github.com/wealdtech/go-ens/commit/b42e4a8) per [ENS deployments](https://docs.ens.domains/learn/deployments/). `NewUniversalResolver` checks `backend.ChainID()` when available; if the chain isn't in the known list (mainnet `1`, Sepolia `11155111`, Holesky `17000`), it returns a typed `UnknownChainError` pointing callers at `NewUniversalResolverAt`. Doc comment updated accordingly. Backends without `ChainID()` keep the previous behaviour. Unit tests in `universalresolver_test.go` (`TestNewUniversalResolver_AcceptsMainnet`, `_AcceptsSepolia`, `_AcceptsHolesky`, `_RejectsUnknownChain`).

File: urerrors.go (hunk 9 above)
Comment (yashgo0018) [reply urerrors.go]: Renamed to `ResolverRevertError` in [`3f49037`](https://github.com/wealdtech/go-ens/commit/3f49037)

File: urerrors_test.go
Comment (yashgo0018) [reply urerrors_test.go:1]: Not intentional - added in [`3f49037`](https://github.com/wealdtech/go-ens/commit/3f49037). `urerrors_test.go` now covers:

- `ResolverError(bytes)` → `ResolverRevertError`
- `HttpError(uint16,string)` → `HTTPGatewayError`
- `ReverseAddressMismatch(string,bytes)` → `ReverseAddressMismatchError`

Each test packs an authentic ABI payload, runs it through `translateURRevert`, and checks the typed error + message.

File: dnsencode.go
Comment (yashgo0018) [reply dnsencode.go:1]: Added in [`46e674e`](https://github.com/wealdtech/go-ens/commit/46e674e) - 22 tests in [`dnsencode_test.go`](https://github.com/wealdtech/go-ens/blob/ensv2-upgrade/dnsencode_test.go) covering normalization, trailing-dot trim, empty labels, non-ASCII/IDN labels, and leading-dot input.

One note on the boundary case: per [ENSIP-10](https://docs.ens.domains/ensip/10), DNS labels are capped at **63 bytes** (RFC 1035 §3.1), not 255, so the tests assert 63/64-byte label boundaries rather than 255/256.

File: universalresolver_test.go (hunk 7 above)
Comment (AntiD2ta) [reply universalresolver_test.go]: If not addressed here, we prefer, at least, having an issue mentioning the missed support for `-short` for the pre-existing legacy live-network tests.

File: dnsencode_test.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+// ---------------------------------------------------------------------------
+// Label-length boundaries.
+//
+// RFC 1035 §3.1 caps a DNS label at 63 bytes. The top 2 bits of the length
+// byte are reserved for compression-pointer flags, so a 64+ byte label cannot
+// be encoded validly.
+//
+// The current implementation uses a 255-byte cap (matching its doc comment).
```
Comment (AntiD2ta) [inline dnsencode_test.go]: Isn't this more accurate?
```suggestion
// The current implementation uses a 63-byte cap.
```

File: universalresolver.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+// NewUniversalResolver returns a UniversalResolver bound to the canonical
+// proxy address on whichever chain the backend is connected to.
+//
+// When the backend exposes a chain ID (ethclient.Client and most production
+// backends do) this verifies the chain is one where the canonical UR is
+// deployed and returns *UnknownChainError otherwise. Backends that don't
+// expose a chain ID skip the check and the canonical address is used as-is —
+// callers on devnets / forks / alt-L1s should prefer NewUniversalResolverAt.
+func NewUniversalResolver(backend bind.ContractBackend) (*UniversalResolver, error) {
```
Comment (AntiD2ta) [inline universalresolver.go]: Related to https://github.com/wealdtech/go-ens/pull/49/changes#r3298127263, we are relying on `context.Background()` here and ignoring any cancellation from the caller's context (such as `resolveName` own `context`).

File: universalresolver_test.go (hunk 7 above)
Comment (yashgo0018) [reply universalresolver_test.go]: created the issue #52

File: dnsencode_test.go (hunk 10 above)
Comment (yashgo0018) [reply dnsencode_test.go]: yes, updated the comments in https://github.com/wealdtech/go-ens/pull/49/changes/88946cd1a13f34815625f6b7fb228efbdf39b5d0

File: universalresolver.go (hunk 11 above)
Comment (yashgo0018) [reply universalresolver.go]: fixed in commit https://github.com/wealdtech/go-ens/pull/49/commits/499727704ee9173395610aa8d9d2e582e653f25b

**Comment by tomdao-tw** [comment]
Hey guys, any blockers to get this merged? We're looking for this too

**Comment by yashgo0018** [comment]
Closes #52 with the commit [1d5f74d](https://github.com/wealdtech/go-ens/pull/49/changes/1d5f74dab6f145261f04812c4febc70a65c23c16)

**Comment by yashgo0018** [comment]
Closes #51 with the commit [3939917](https://github.com/wealdtech/go-ens/pull/49/commits/3939917c4fa374be0a511d6446a18604b0a07da8)
